File: src/Application/Client/Java/ByteKeyIterator.java
package com.scalien.scaliendb;

import java.util.List;
import java.util.Collections;

public class ByteKeyIterator implements java.lang.Iterable<byte[]>, java.util.Iterator<byte[]>
{
    private Table table;
    private byte[] startKey;
    private byte[] endKey;
    private byte[] prefix;
    private int count;
    private int gran = 100;
    private List<byte[]> keys;
    private int pos;
 
    public ByteKeyIterator(Table table, ByteRangeParams ps) throws SDBPException {
        this.table = table;
        this.startKey = ps.startKey;
        this.endKey = ps.endKey;
        this.prefix = ps.prefix;
        this.count = ps.count;
        
        query(false);
    }

    // for Iterable
    public java.util.Iterator<byte[]> iterator() {
        return this;
    }
    
    // for Iterator
    public boolean hasNext() {
        try {
            if (count == 0)
                return false;
            if (pos == keys.size()) {
                if (keys.size() < gran)
                    return false;
                startKey = keys.get(keys.size()-1);
                query(true);
            }
            if (keys.size() == 0)
                return false;
            return true;
        }
        catch(SDBPException e) {
            return false;
        }
    }
    
    // for Iterator
    public byte[] next() {
        byte[] e = keys.get(pos);
        pos++;
        if (count > 0)
            count--;
        return e;
    }
    
    // for Iterator
    public void remove() throws UnsupportedOperationException {
        throw new UnsupportedOperationException();
    }
    
    private void query(boolean skip) throws SDBPException {
        int num;

        num = gran;
        if (count > 0 && count < gran)
            num = count;

        keys = table.getClient().listKeys(table.getTableID(), startKey, endKey, prefix, num, skip);
        Collections.sort(keys, new ByteArrayComparator());
        pos = 0;
    }
}

File: src/Application/Client/CSharp/ScalienClient/ByteKeyIterator.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace Scalien
{
    public class ByteKeyIterator : IEnumerable<byte[]>, IEnumerator<byte[]>
    {
        Table table;
        bool forwardDirection;
        byte[] startKey;
        byte[] endKey;
        byte[] prefix;
        long count;
        uint granularity = 100;
        int pos;
        List<byte[]> keys;

        public ByteKeyIterator(Table table, ByteRangeParams ps)
        {
            this.table = table;
            this.startKey = ps.startKey;
            this.endKey = ps.endKey;
            this.prefix = ps.prefix;
            this.count = ps.count;
            this.granularity = ps.granularity;
            this.forwardDirection = ps.forwardDirection;
            Query(false);
        }

        private void Query(bool skip)
        {
            uint num;

            num = granularity;
            if (count > 0 && count < granularity)
                num = (uint)count;

            keys = table.Client.ListKeys(table.TableID, startKey, endKey, prefix, num, forwardDirection, skip);
            pos = 0;
        }

        #region IEnumerable interface

        IEnumerator IEnumerable.GetEnumerator()
        {
            return (IEnumerator)GetEnumerator();
        }

        public IEnumerator<byte[]> GetEnumerator()
        {
            return this;
        }

        #endregion

        #region IEnumerator interface

        public bool MoveNext()
        {
            if (count == 0)
                return false;
            if (pos == keys.Count)
            {
                if (keys.Count < granularity)
                    return false;
                startKey = keys[keys.Count - 1];
                Query(true);
            }
            if (keys.Count == 0)
                return false;

            pos++;
            if (count > 0)
                count--;
            return true;
        }

        public void Reset()
        {
            throw new NotSupportedException();
        }

        public byte[] Current
        {
            get
            {
                return keys[pos - 1];
            }
        }

        object IEnumerator.Current
        {
            get
            {
                return Current;
            }
        }

        void IDisposable.Dispose()
        {
        }

        #endregion
    }
}
